Clean up official mappings and other related fixes

SizableShrimp opened the pull request:
This PR fixes some issues with official mappings and adds greater support for them including information about constructors and being able to use snapshot information.

* Support snapshots with official mappings and hardcode the first versions supported with official mappings so users cannot query invalid versions before official mappings were released
* Properly store information of unmapped data in official mappings like <init> and other methods required to have the same name by external libs inheritance
* Fix command hint to use `getName()`
* Fix `CommandMappings` sometimes replying incorrectly when updating guild settings with a flag
* Add `defaultStable` flag which is set to true in official mappings so that by default 1.16.5 is used instead of the latest snapshot when the field is unset for a guild.
* Fix manifest version info for official mappings
* Refactor display of official mappings and add more info like parameters and method descriptor

SizableShrimp committed:
Clean up official mappings and other fixes

Support snapshots with official mappings and hardcode the versions supported with official mappings
Properly store information of unmapped data in official mappings like <init>
Fix command hint to use `getName()`
Fix `CommandMappings` sometimes replying incorrectly when updating guild settings with a flag
Add `defaultStable` flag which is set to true in official mappings
Fix manifest version info for official mappings
Refactor display of official mappings and add more info

tterrag1098 commented on src/main/java/com/tterrag/k9/commands/CommandMappings.java:
```diff
@@ -122,18 +127,18 @@ public void save(File dataFolder, Gson gson) {
     }
 
-    protected Flux<? extends Mapping> findMappings(CommandContext ctx) {        
+    protected Mono<Optional<List<Mapping>>> findMappings(CommandContext ctx) {
```
This is super pointless. `Mono` can be empty, remove the `Optional`.

tterrag1098 commented on src/main/java/com/tterrag/k9/mappings/official/OfficialMapping.java:
```diff
@@ -116,7 +133,12 @@ public String getDesc(NameType name) {
         if (intermediate == null) {
-            if (type == MappingType.CLASS) {
+            // Should help with random issues
+            mappedDesc.clear();
```
This is not a good idea, what's the point of the cache then?

SizableShrimp replied:
The cache is cleared when resolving getIntermediate to fix lingering issues. After that, it stays cached. This is only called once when the intermediate is first requested.

tterrag1098 commented on src/main/java/com/tterrag/k9/mappings/official/OfficialMapping.java:
```diff
@@ -49,47 +49,64 @@
         }
 
         if (getMemberClass() != null)
-            builder.append("\n__Type__: `").append(getMemberClass()).append("`");
```
I don't see a replacement for this linebreak?

tterrag1098 commented on src/main/java/com/tterrag/k9/mappings/official/OfficialMapping.java:
```diff
@@ -127,6 +149,11 @@ public String getIntermediate() {
             }
+
+            // Special cases like <init> and other stuff required to have the same name by external libs
+            if (srgs != null && intermediate.isEmpty() && original.equals(name)) {
```
This "is special" check is done multiple times, move it to a method.

tterrag1098 commented on src/main/java/com/tterrag/k9/mappings/official/ManifestJson.java:
```diff
@@ -27,8 +30,13 @@
 
     @Data
     @Setter(AccessLevel.NONE)
+    @EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
Why not just put `@EqualsAndHashCode.Exclude` on `url`?

SizableShrimp committed:
Fixes

Fix useless Optional boxing
Fix missing newline separator
Extract out a method to check whether a mapping is "special"
Use `@EqualsAndHashCode.Exclude` instead

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>
